src/openssl/mod.rs

```diff
@@ -0,0 +1,987 @@
+    /// (as `BigNum`)
+    ///
+    /// # Arguments
+    ///
+    /// * `pi`  - A slice of octets representing the VRF proof
+    ///
+    /// # Returns
+    ///
+    /// * A tuple containing the gamma point, and parameters `c` and `s`
+    fn decode_proof(&mut self, pi: &[u8]) -> Result<(EcPoint, BigNum, BigNum), Error> {
+        let gamma_oct = if self.qlen % 8 > 0 {
+            self.qlen / 8 + 2
+        } else {
+            self.qlen / 8 + 1
+        };
+        let c_oct = if self.n % 8 > 0 {
+            self.n / 8 + 1
+        } else {
+            self.n / 8
+        };
+
+        if pi.len() * 8 < gamma_oct + c_oct * 3 {
+            return Err(Error::InvalidPiLength);
+        }
+        let gamma_point = EcPoint::from_bytes(&self.group, &pi[0..gamma_oct], &mut self.bn_ctx)?;
+        let c = BigNum::from_slice(&pi[gamma_oct..gamma_oct + c_oct])?;
+        let s = BigNum::from_slice(&pi[gamma_oct + c_oct..])?;
+
+        Ok((gamma_point, c, s))
+    }
+
+    /// Computes the VRF hash output as result of the digest of a ciphersuite-dependent prefix
+    /// concatenated with the gamma point (vrf-draft-04, section 5.2)
+    ///
+    /// # Arguments
+    ///
+    /// * `gamma`  - A EC point representing the VRF gamma
+    ///
+    /// # Returns
+    ///
+    /// * A vector of octets with the VRF hash output
+    fn proof_to_hash(&mut self, gamma: &EcPoint) -> Result<Vec<u8>, Error> {
+        // Multiply gamma with cofactor
+        let mut gamma_cof = EcPoint::new(&self.group.as_ref())?;
+        gamma_cof.mul(
+            &self.group.as_ref(),
+            &gamma,
+            &BigNum::from_slice(&[self.cofactor])?.as_ref(),
+            &self.bn_ctx,
+        )?;
+
+        let gamma_string = gamma_cof.to_bytes(
+            &self.group,
+            PointConversionForm::COMPRESSED,
+            &mut self.bn_ctx,
+        )?;
+
+        let hash = hash(
+            self.hasher,
+            &[
+                &[self.cipher_suite.suite_string()],
+                &[0x03],
+                &gamma_string[..],
+            ]
+            .concat(),
+        )
+        .map(|hash| hash.to_vec())?;
+
+        Ok(hash)
+    }
+}
+
+/// VRFs are objects capable of generating and verifying proofs
+impl VRF<&[u8], &[u8]> for ECVRF {
+    type Error = Error;
+
+    /// Generates proof from a secret key and message as specified in the VRF-draft-04
+    /// (vrf-draft-04, section 5.1)
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A slice representing the secret key in octets
+    /// * `alpha` - A slice representing the message in octets
+    ///
+    /// # Returns
+    ///
+    /// * If successful, a vector of octets representing the proof of the VRF
+    fn prove(&mut self, x: &[u8], alpha: &[u8]) -> Result<Vec<u8>, Error> {
+        // Step 1: derive public key from secret key
+        // Y = x * B
+        //TODO: validate secret key length?
```

Comment:
This validation wouldn't be necessary if the secret key was a new type `&SecretKey` instead of `&[u8]`.

Reply:
Yet if `SecretKey` is a type alias for an array of known length, is up to the implementor to define its length, as different curves could use different secret key lengths. Amirite @girazoki @mariocao ?

Reply:
Yep